### ISSUE TITLE
Memory management of class clusters

### ISSUE THREAD

**Issue by ivanbuhov** [NativeScript/ios-runtime#214]
In the current implementation when an interface is instantiated we increase the `refCount` of the pointer returned from `alloc`. When a class cluster is instantiated, the pointer returned from `alloc` is not the same as the pointer returned from the initiliazer method (e.g. `init`), so we increase the `refCount` of the pointer returned from `alloc` and in the initializer the placeholder pointed by this pointer is deallocated, and new pointer is returned. We never call `retain` of the newly allocated object but we call its `release` method before deallocation which leads to over releasing.

The following code crashes the app with EXC_BAD_ACCESS:
```
var fd1 = UIFontDescriptor.alloc().init();
var fd2 = fd1.fontDescriptorWithSymbolicTraits(0);
```
Possible workaround is to use `new ClassCluster(args)` but it works only if there is no initializers collision (initializers with the same signature).

Possible solution is to provide a custom `alloc` implementation which calls the original `alloc` method, wraps the result in `PointerInstance` and sets its prototype to be the prototype of the instantiated class. We must no more retain the returned result of `alloc`. All initializers should be marked in metadata to return a retained object. We will `retain` the result of the initilizer method call.

* [x]  <del>Update metadata. Change all initiliazers to return retained object.</del>
* [x] Update metadata. Change `alloc` method to return not retained object.
* [x] Update metadata. Mark with flag all initializer methods.
* [x] Introduce the new `alloc` method.
* [x] Introduce AllocedObject.
* [x] Fix `instancetype` marshalling.
* [x] Introduce a property to `ObjCSuperObject`, `ObjCWrapperObject`, `ObjCConstructorBase` where to hold their instances `Structure`.
 
Related issues: https://github.com/NativeScript/ios-runtime/issues/152

**Comment by ivanbuhov** [comment]
We will introduce an `AllocedObject` class which wraps an object returrned from `alloc` message sent to a `ObjCConstructorNative`. So when an initializer is called to `AllocedObject` instance we will release  the result to balance the `refCount`. This change will break the marshalling of `instancetype`, because it retrives the structure of the returned `instancetype` from the `this` object the method is called on, so we have to handle it.

**Comment by ivanbuhov** [comment]
Fixed with 56f762d08ab733d1763596265daa2696d8fd3f92
